contracts/cw-proposal-single/src/contract.rs

```diff
@@ -609,9 +609,47 @@ pub fn query(deps: Deps, env: Env, msg: QueryMsg) -> StdResult<Binary> {
         } => query_reverse_proposals(deps, env, start_before, limit),
         QueryMsg::ProposalHooks {} => to_binary(&PROPOSAL_HOOKS.query_hooks(deps)?),
         QueryMsg::VoteHooks {} => to_binary(&VOTE_HOOKS.query_hooks(deps)?),
+        QueryMsg::FindProposals {
+            wallet,
+            status,
+            wallet_vote,
+            start_after,
+            limit,
+        } => query_find_proposals(deps, env, wallet, status, wallet_vote, start_after, limit),
     }
 }
 
+fn query_find_proposals(
+    deps: Deps,
+    env: Env,
+    wallet: String,
+    status: Option<Status>,
+    wallet_vote: Option<Vote>,
+    start_after: Option<u64>,
+    limit: Option<u64>,
+) -> StdResult<Binary> {
+    let wallet: Addr = deps.api.addr_validate(&wallet)?;
+    let min = start_after.map(Bound::exclusive);
+    let limit = limit.unwrap_or(DEFAULT_LIMIT);
+    let props: Vec<ProposalResponse> = PROPOSALS
+        .range(deps.storage, min, None, Order::Ascending)
+        .take(limit as usize)
+        .collect::<StdResult<Vec<(u64, SingleChoiceProposal)>>>()?
+        .into_iter()
+        .filter_map(|(p_id, prop)| {
+            if status.map_or(true, |st| st == prop.status) {
+                if let Ok(Some(ballot)) = BALLOTS.may_load(deps.storage, (p_id, wallet.clone())) {
+                    if wallet_vote.map_or(true, |v| v == ballot.vote) {
+                        return Some(prop.into_response(&env.block, p_id));
+                    }
+                }
+            }
+            None
+        })
```

Comment:
I'm not entirely sure this works.

Say I have proposals (1), (2), and (3) and I have voted no proposals (1) and (2). If I run this query to find "the first one" (`limit == 1`) I have not voted on, the `limit` call here will cause us to only check the first proposal and I think I'll get an empty list back.

I think fixing this is a pretty hard problem if we want to be gas efficient. My strong suspicion is that we'll need additional state.

Reply:
One solution that would likely be fine would just be to move the limit call to after `filter_map`. Queries very rarely run out of gas because of compute limits. I suspect we could have thousands of proposals before this ran out of gas.

Another would be to use your suggestion of the `proposals_by_addr` map. Would love to talk through this in the Discord! https://github.com/DA0-DA0/dao-contracts/issues/237#issuecomment-1187977788

Reply:
My idea was basically to make `FindProposals`, but sorted, maybe we could rename it to smth like `FilterListProposals` then, what you think?(not sure about new naming for `limit`) 

And I think `start_after` can become pretty user-unfriendly, let's say user put some `start_after` do we `start_after` `proposal_id` or index of user proposals. 

But maybe I'm just **too worried** about the gas and no one will have enough proposals for it to break :DD

Reply:
Great point about `start_after` here. I like your idea of making the query `FilterListProposals`, maybe in order to make `start_after` play nice, we could return a response like this:

```
struct FilterListProposalsResponse {
  proposals: Vec<Proposal>,
  last_checked: String
}
```

Then, a frontend can use that `last_checked` value to set `start_after` for the next iteration. I think that if we do this we can have pagination (no gas issues!) and get the queries we want.